Updates and cleanup for some configuration-related items (2.6.x)

JCgH4164838Gh792C124B5 opened the pull request:
Updates and cleanup for some configuration-related items.
- Made several attributes final.
- Added limiter logic to the setting of certain configuration elements, and increased logging around them.
- Updated unit tests impacted by the limiter logic, including minor typo fixes and locale consistency.
- Minor cleanup while making the changes.

aleksandr-m commented on core/src/main/java/com/opensymphony/xwork2/config/providers/XWorkConfigurationProvider.java:
```diff
@@ -224,6 +224,7 @@ public void register(ContainerBuilder builder, LocatableProperties props)
         props.setProperty(StrutsConstants.STRUTS_ENABLE_OGNL_EXPRESSION_CACHE, Boolean.TRUE.toString());
         props.setProperty(StrutsConstants.STRUTS_ENABLE_OGNL_EVAL_EXPRESSION, Boolean.FALSE.toString());
         props.setProperty(StrutsConstants.STRUTS_CONFIGURATION_XML_RELOAD, Boolean.FALSE.toString());
+        props.setProperty(StrutsConstants.STRUTS_ALLOW_STATIC_METHOD_ACCESS, Boolean.FALSE.toString());
```
Why? Is it not working?

JCgH4164838Gh792C124B5 replied:
Hello @aleksandr-m .
Are you asking why the new entry on line 227 ?  Simply to ensure the props includes a set property for that element, in line with the other significant ones already present.

aleksandr-m replied:
But there is more constants. Why to add especially this one?

JCgH4164838Gh792C124B5 replied:
Properties like devMode, DMI, eval_expression are all "privilege-related" (and set to false), just like static method access.  Since the others were present, it seemed that static method access should be there too (for completeness, also set to false).
That was my reasoning anyway.  :)

aleksandr-m commented on core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java:
```diff
@@ -82,7 +86,18 @@ public void setXWorkConverter(XWorkConverter conv) {
     @Inject(StrutsConstants.STRUTS_DEVMODE)
     public void setDevMode(String mode) {
-        this.devMode = BooleanUtils.toBoolean(mode);
+        final boolean booleanMode = BooleanUtils.toBoolean(mode);
```
Why? If I remember correctly then injection can happen multiple times and some plugins, for example, can override already injected properties. That is how some things work (`disallowProxyMemberAccess` is one of this things).

yasserzamani replied:
@apache/struts-committers , I think here @JCgH4164838Gh792C124B5 is trying to prevent hacker from the ability of resetting those values that menaces overall security. Whether it works or breaks here, In general it seems a nice devised practice :+1: thank you!

aleksandr-m replied:
@yasserzamani How it is helpful if basic stuff isn't working?

If this is really the case then they can be protected in some other way.

yasserzamani replied:
Yeah basic stuff should work. I actually meant the idea itself - to prevent hacker from the ability of resetting those values that menaces overall security.

JCgH4164838Gh792C124B5 replied:
Hello Struts Team.

The primary intention for this PR was only to try and limit the ability to change certain elements (such as devMode), for which there didn't seem to be a good reason for them to change after the 1st (initial) configuration.  

It was more aimed at avoiding inadvertent/accidental state changes for important state than anything else.

None of the regression tests, nor interaction with the Showcase application encountered any issues so it seemed like a reasonable approach to take.

If there are certain properties (such as `disallowProxyMemberAccess`) for which subsequent changes are necessary/valid, then they could be omitted by this or another mechanism that attempts to limit such changes.

JCgH4164838Gh792C124B5 replied:
Hello @aleksandr-m  and @yasserzamani .  

Do either of you think _any_ of the properties I attempted to apply change-limiters to could/should have the change-limiter applied to them ?

I could eliminate some or all of the change-limiters in a revision of this PR (and PR #291) that only retains the additional logging around property changes.  Please let me know what you think.

aleksandr-m replied:
> Do either of you think any of the properties I attempted to apply change-limiters to could/should have the change-limiter applied to them ?

Well, currently I don't see what benefits does it bring.

yasserzamani replied:
Hi @JCgH4164838Gh792C124B5 , I like your idea but to be frank, I'm not able to have a vision on future side effects.

But I remembered one of my old ideas that not only has somehow same benefits but also is lower risky and is more readable: **Refactoring all methods annotated with `@com.opensymphony.xwork2.inject.Inject` from `public setXXX` to `private injectXXX`.** This preserves those secure against OGNL access. As a glimpse, with IntelliJ IDE's refactoring tools, I did it on `OgnlUtil.setDevMode` and worked  :)

aleksandr-m commented on core/src/main/resources/struts-default.xml:
```diff
@@ -61,6 +61,7 @@
                 freemarker.ext.rhino.,
                 sun.reflect.,
                 javassist.,
+                org.objectweb.asm.,
```
This must be in separate commit / PR.

aleksandr-m commented on core/src/test/java/com/opensymphony/xwork2/ognl/OgnlUtilTest.java:
```diff
@@ -375,7 +381,7 @@ public void testSetPropertiesDate() {
 
         //UK style test
-        cal = Calendar.getInstance();
+        cal = Calendar.getInstance(Locale.UK);
```
Why?

JCgH4164838Gh792C124B5 replied:
Hello @aleksandr-m .

I found during debugging within an IDE that the `OgnlUtilTest` test `testSetPropertiesDate` was failing (even though it worked when run in the Maven build - on Windows in both cases).  It might be a combination of OS/environment conditions for the default Java locale - so I explicitly set the locale values (based on the existing comments) so that the test worked consistently both inside and outside the Maven build.

aleksandr-m replied:
Ok.

aleksandr-m commented on core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java:
```diff
@@ -57,17 +57,21 @@
     private Container container;
-    private boolean allowStaticMethodAccess;
-    private boolean disallowProxyMemberAccess;
+    private boolean allowStaticMethodAccess;            // Defaults to false
```
No need for the comments here.

aleksandr-m replied:
Don't be. :) Unnecessary comments just clutter code and can take focus away from the important stuff.

yasserzamani commented on core/src/main/java/com/opensymphony/xwork2/security/DefaultExcludedPatternsChecker.java:
```diff
@@ -66,9 +76,14 @@ public void setAdditionalExcludePatterns(String excludePatterns) {
     public void setDynamicMethodInvocation(String dmiValue) {
         if (!BooleanUtils.toBoolean(dmiValue)) {
-            LOG.debug("DMI is disabled, adding DMI related excluded patterns");
+            if (LOG.isDebugEnabled()) {
```
Please revise all `if (LOG...` statements in this PR according to [this answer](https://stackoverflow.com/a/29561054/1362623).

yasserzamani commented on core/src/main/java/com/opensymphony/xwork2/ognl/OgnlUtil.java:
```diff
@@ -92,10 +107,17 @@ public void setEnableExpressionCache(String cache) {
-        if(enableEvalExpression){
-            LOG.warn("Enabling OGNL expression evaluation may introduce security risks " +
-                    "(see http://struts.apache.org/release/2.3.x/docs/s2-013.html for further details)");
+        if (!this.enableEvalExpressionValueSet) {
```
Sorry you did a lot of work but please replace all change-limiters logic including their unit test with proposed solution above. Thanks in advance!

JCgH4164838Gh792C124B5 commented:
Hello @yasserzamani  and @aleksandr-m .

This PR has been updated with changes that hopefully addresses most of the concerns that were provided in your feedback.

The limiter logic was removed and instead of the full-refactoring of the injectable methods it only reduces the access level of most of the relevant methods to `protected` (improves access safety with less chance of impacting plugins or other elements negatively).  It's not the level of refactoring that @yasserzamani  suggested above, but is _a small step in that direction_ (although slightly less risky in terms of potential impacts).

Let me know if the changes in the PR look better now.

yasserzamani commented on core/src/main/java/com/opensymphony/xwork2/security/DefaultAcceptedPatternsChecker.java:
```diff
@@ -70,7 +70,14 @@ public void setAcceptedPatterns(String[] additionalPatterns) {
     public void setAcceptedPatterns(Set<String> patterns) {
-        LOG.trace("Sets accepted patterns [{}]", patterns);
+        if (acceptedPatterns == null) {
+            LOG.debug("Replacing accepted patterns [{}] with [{}], be aware that this affects all instances and safety of your application!",
```
It should be `LOG.debug("Sets accepted patterns [{}]", patterns);`

yasserzamani commented on core/src/main/java/com/opensymphony/xwork2/security/DefaultExcludedPatternsChecker.java:
```diff
@@ -46,9 +46,15 @@ public DefaultExcludedPatternsChecker() {
+                        excludedPatterns, excludePatterns);
+        }
+        else {
+            LOG.debug("Overriding excluded patterns [{}] with [{}], be aware that this affects all instances and safety of your application!",
```
It should be `LOG.debug("Overriding excluded patterns with [{}]", excludePatterns);`

yasserzamani commented on core/src/main/java/com/opensymphony/xwork2/security/DefaultExcludedPatternsChecker.java:
```diff
@@ -80,7 +89,14 @@ public void setExcludedPatterns(String[] patterns) {
+                        excludedPatterns, patterns);
+        }
+        else {
+            LOG.debug("Replacing excluded patterns [{}] with [{}], be aware that this affects all instances and safety of your application!",
```
It should be `LOG.debug("Sets excluded patterns [{}]", patterns);`

JCgH4164838Gh792C124B5 replied:
Hello @yasserzamani .
The two `debug` log statements have been updated in the most recent commit based on your comments/request.  I left in the extra warning text information but the wording and parameters of the start of the log output match what you requested.  :)

yasserzamani replied:
Yes it's great thank you :)

@aleksandr-m , if you consent please approve your review, then we're OK to squash and merge this one :)

aleksandr-m replied:
@yasserzamani Done. Squashing would be great. ;)
